letta/agent.py

```diff
@@ -651,6 +652,48 @@ def _handle_ai_response(
 
                 function_args["self"] = self  # need to attach self to arg since it's dynamically linked
 
+                # Execute tool in sandbox
+                if (function_name in ["print_hello_world", "print_message"]): # remove hardcode
+                    sbx = Sandbox(api_key="e2b_99c3fa2617e66233a57bf267e1e1f462745b0da5") # store api key in env var
+                    code = ""
+                    
+                    # 1. Import dependencies
+                    # package = ""
+                    # sbx.commands.run(f"pip3 install {package}")
+
+                    # 2. Set Params
+                    for param in function_args:
+                        if param != "self":
+                            code += param + ' = "' + function_args[param] + '"\n'
+
+                    # 3. Add function source code
+                    tool = [tool for tool in self.tools if tool.name == function_name][0]
+                    code += tool.source_code + "\n"
+
+                    # 4. Add function call
+                    code += function_name + "("
+
+                    # 5. Populate params for function call
+                    for param in function_args:
+                        if param != "self":
+                            code += param + ","
+                    
+                    code += ")"
+                    
+                    # 6. Execute code and extract result or throw
+                    execution = sbx.run_code(code)
+                    if execution.error is not None:
+                        raise execution.error
+                    elif len(execution.results) == 0:
+                        function_response = ""
+                    else:
+                        function_response = execution.results[0].text
```

Comment:
is there any issue with typing here - e.g. if the function returns a list? does this assume the function is always returning a string?

Reply:
e2b will return lists as a string and store it in the text field! i.e.
```
python3
>>> from e2b_code_interpreter import Sandbox
>>> sbx = Sandbox(api_key=<API_KEY>)
>>> e = sbx.run_code("""
... def func(a):
...     return [1] * a
... 
... func(5)
... """)
>>> e
Execution(Results: [Result([1, 1, 1, 1, 1])], Logs: Logs(stdout: [], stderr: []), Error: None)
>>> e.results[0].text
'[1, 1, 1, 1, 1]'
```
It seems like the primitive types will show up here, but there are more complex types that have their own [field](https://github.com/e2b-dev/code-interpreter/blob/main/python/e2b_code_interpreter/models.py#L94-L122). Should be straightforward to extend as needed in the future.